Form data support

Michae1Weiss opened the pull request:
Hi everyone!

This PR shows an example of support for uploading files via Swagger.

Usage example with ```werkzeug.datastructures.FileStorage``` (Flask):

```python3
from werkzeug.datastructures import FileStorage
from pydantic import BaseModel
from spectree import models

class File(BaseModel):
    uid: str
    file: models.BaseFile

@api.validate(form_data=File)
def with_file():
    file_dict = request.context.form_data.file.dict()
    file = FileStorage(**file_dict)
    return jsonify()
...
```

## Issues:
1. If uploading files using the Falcon framework plugin - ```content_length``` is always 0.

## Questions:
1. How to get ```content_length``` from ```falcon.cyutil.reader.BufferedReader``` object?
2. Initially ```SpecTree.validate``` only accepts ```json``` parameter to define request body. Would it be correct to separate request body schemas for ```multipart/form-data``` and ```application/json``` content types with different parameters (```json``` and ```from_data```) like I did or is there a cleaner way to do it?</br>
For example, the 'drf-yasg' module uses a generic ```request_body``` parameter for both content types.
To define the content type **'drf-yasg'** parses the view class object and if a view include only form parsers - swagger will support file upload.
[**drf-yasg implementation**](https://github.com/axnsan12/drf-yasg/blob/master/src/drf_yasg/utils.py#L366)</br>
Wouldn't it make sense to make a generic ```request_body``` parameter like in **'drf-yasg**' and understand expected request content type of the request body by some flag for example?

kemingy commented on spectree/plugins/falcon_plugin.py:
```diff
@@ -182,12 +182,32 @@ def request_validation(self, req, query, json, headers, cookies):
+                        "content_type": part.content_type,
+                        "stream": part.stream.read(),
+                    }
+                # TODO: add support for other media types?
+                # TODO: OR handle unsupported media types?
```
Refs: https://www.iana.org/assignments/media-types/media-types.xhtml

How about the following?

* font
* video
* message
* model

I wonder what's the major difference among these types for file handlers?

kemingy commented on spectree/plugins/falcon_plugin.py:
```diff
@@ -182,12 +182,32 @@ def request_validation(self, req, query, json, headers, cookies):
+                    data[part.name] = {
+                        "filename": part.filename,
+                        "name": part.name,
+                        # "content_length": len(part.data),  # FIXME: raises error if file is too big
+                        # sadly, BufferedReader don't have 'tell' method
+                        "content_length": 0,  # TODO: replace placeholder with something that will work
+                        "content_type": part.content_type,
+                        "stream": part.stream.read(),
```
Since we already consume the stream buffer, why not just use `part.data` to access the data and content length?

I'm not sure if it's possible to keep the buffer and let the user consume it later.

kemingy commented on spectree/plugins/falcon_plugin.py:
```diff
@@ -182,12 +182,32 @@ def request_validation(self, req, query, json, headers, cookies):
             req.context.json = json.parse_obj(media)
+        elif form_data and media is not None:
+            data = {}
+            for part in media:  # TODO: add support for falcon 2.0 (media is always None for 'multipart/form-data')
```
BTW, async falcon should also be in the TODO list.

kemingy commented on spectree/plugins/starlette_plugin.py:
```diff
@@ -37,10 +38,28 @@ def register_route(self, app):
         request.context = Context(
             query.parse_obj(request.query_params) if query else None,
             json.parse_raw(await request.body() or "{}") if json else None,
+            form_data.parse_obj(data or "{}") if form_data and data else None,
```
```suggestion
            form_data.parse_obj(data or {}) if form_data and data else None,
```

kemingy commented:
Thanks for your great work!

You can fix the lint by running `make format` and test with `make lint`.

Michae1Weiss commented:
@kemingy Thanks for review! I’ve added form-data support for async falcon and implemented various optimizations and fixes based on your comments. It's now time to write some tests for it :)
I will try to write them next week.
I'd be grateful for a second review and feedback. I'm not quite sure about some of the fixes (still need to learn more about asgi and cgi)

kemingy commented:
Hi, @Michae1Weiss. Thanks for your contribution. It may take me some time to review the code. I will try to give you some feedback by the end of this week.

yihong0618 commented on spectree/plugins/starlette_plugin.py:
```diff
@@ -38,9 +39,27 @@ def register_route(self, app):
 
     async def request_validation(self, request, query, json, form_data, headers, cookies):
+        data = {}
+        if request.headers.get('content-type', '').split(";")[0] in self.FORM_MIMETYPE:
```
if get the default `''` and split here index[0]  will raise an error

Michae1Weiss replied:
Thanks for the feedback!
You're right, splitting an empty string with a `None` separator returns [] and `''.split()[0]` will raise an error.
But in this case `split()` uses a separator `;`. So `''.split(';')[0]` will return an empty string
From Python documentation:
`Splitting an empty string with a specified separator returns [''].`
Refs: https://docs.python.org/3/library/stdtypes.html#str.split

yihong0618 replied:
got your point, you are right, cool~

yihong0618 commented on spectree/plugins/flask_plugin.py:
```diff
@@ -122,30 +122,34 @@ def parse_path(self, route, path_parameter_descriptions):
+                form_dict.update(**request.files.to_dict())
+                if form:
+                    req_form = form_dict
+                elif json:
```
maybe if here?

yedpodtrzitko replied:
elsewhere in this PR there is a condition that will raise exception if both `json` and `form` exist, so it seems like they are mutually exclusive

kemingy commented on spectree/plugins/falcon_plugin.py:
```diff
@@ -167,27 +170,59 @@ def parse_path(self, route, path_parameter_descriptions):
+
+            try:
+                _form = cgi.FieldStorage(fp=req.stream, environ=req.env)  # TODO: rename
+            except ValueError:  # Invalid boundary?
```
Will [`falcon.Request.bounded_stream`](https://falcon.readthedocs.io/en/stable/api/request_and_response_wsgi.html?highlight=file#falcon.Request.bounded_stream) be better to avoid this exception?

kemingy commented on spectree/plugins/falcon_plugin.py:
```diff
@@ -167,27 +170,59 @@ def parse_path(self, route, path_parameter_descriptions):
+            encoding, filename = encoded.split("''")
+            field.filename = filename
+            field.file = BytesIO(field.file.read().encode(encoding))
+        if getattr(field, "filename", False):
```
I think `getattr(field, "filename", False)` will always return `None` since the type of `field` is always `cgi.FieldStorage`.

kemingy commented on spectree/plugins/falcon_plugin.py:
```diff
@@ -167,27 +170,59 @@ def parse_path(self, route, path_parameter_descriptions):
+            return [self.parse_field(subfield) for subfield in field]
+
+        # When file name isn't ascii FieldStorage will not consider it.
+        encoded = field.disposition_options.get("filename*")
```
Is `filename*` a regex?

kemingy commented on spectree/plugins/falcon_plugin.py:
```diff
@@ -167,27 +170,59 @@ def parse_path(self, route, path_parameter_descriptions):
+        # When file name isn't ascii FieldStorage will not consider it.
+        encoded = field.disposition_options.get("filename*")
+        if encoded:
+            encoding, filename = encoded.split("''")
```
Any example for this one? I'm not sure when will the `field.filename` contains `''`.

kemingy reviewed (commented):
Thanks for your great work. Feel free to leave some comments. Especially for the `cgi.FileStorage` part. I'm not very familiar with it. I'm not sure if I understand your code correctly.

yedpodtrzitko commented on spectree/plugins/flask_plugin.py:
```diff
@@ -122,21 +122,34 @@ def parse_path(self, route, path_parameter_descriptions):
         req_cookies = request.cookies or {}
+
+        if request.mimetype in self.FORM_MIMETYPE:
+            body = form or json or None
```
what is `or None` for?

yedpodtrzitko commented on spectree/plugins/falcon_plugin.py:
```diff
@@ -1,6 +1,9 @@
+import cgi
```
`cgi` is deprecated since Python 3.8 and is removed in Python 3.10

kemingy commented:
* close this because of #225
